cache/cache.go

```diff
@@ -0,0 +1,144 @@
+package cache
+
+import (
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+)
+
+const NodeSize = 32
+
+type Writer struct {
+	*cache
+}
+
+func NewWriterWithLayerFactories(layerFactories []LayerFactory) *Writer {
+	return &Writer{
+		cache: &cache{
+			layers:         make(map[uint]LayerReadWriter),
+			layerFactories: layerFactories,
+		},
+	}
+}
+
+func (c *Writer) GetLayerWriter(layer uint) LayerWriter {
+	layerReadWriter, found := c.layers[layer]
+	if !found {
+		layerReadWriter = c.fillLayerIfNeeded(layer)
+	}
+	return layerReadWriter
+}
+
+func (c *Writer) SetHash(hashFunc func(lChild, rChild []byte) []byte) {
+	c.hash = hashFunc
+}
+
+func (c *Writer) GetReader() (*Reader, error) {
+	err := c.ValidateStructure()
+	if err != nil {
+		return nil, err
+	}
+	return &Reader{c.cache}, nil
+}
+
+type Reader struct {
+	*cache
+}
+
+type cache struct {
+	layers         map[uint]LayerReadWriter
+	hash           func(lChild, rChild []byte) []byte
+	layerFactories []LayerFactory
+}
+
+func (c *Reader) GetLayerReader(layer uint) LayerReader {
+	return c.layers[layer]
+}
+
+func (c *Reader) GetHashFunc() func(lChild, rChild []byte) []byte {
+	return c.hash
+}
+
+func (c *cache) Print(bottom, top int) {
+	for i := top; i >= bottom; i-- {
+		print("| ")
+		sliceReadWriter, ok := c.layers[uint(i)].(*SliceReadWriter)
+		if !ok {
+			println("-- layer is not a SliceReadWriter --")
+			continue
+		}
+		for _, n := range sliceReadWriter.slice {
+			printSpaces(numSpaces(i))
+			fmt.Print(hex.EncodeToString(n[:2]))
+			printSpaces(numSpaces(i))
+		}
+		println(" |")
+	}
+}
+
+func (c *cache) ValidateStructure() error {
+	// Verify we got the base layer.
+	if _, found := c.layers[0]; !found {
+		return errors.New("reader for base layer must be included")
+	}
+	width := c.layers[0].Width()
+	if width == 0 {
+		return errors.New("base layer cannot be empty")
+	}
+	height := RootHeightFromWidth(width)
+	for i := uint(0); i < height; i++ {
+		if _, found := c.layers[i]; found && c.layers[i].Width() != width {
+			return fmt.Errorf("reader at layer %d has width %d instead of %d", i, c.layers[i].Width(), width)
+		}
+		width >>= 1
+	}
+	return nil
+}
+
+func (c *cache) fillLayerIfNeeded(layer uint) LayerReadWriter {
+	for _, factory := range c.layerFactories {
+		layerReadWriter := factory(layer)
+		if layerReadWriter != nil {
+			c.layers[layer] = layerReadWriter
+			return layerReadWriter
+		}
+	}
+	return nil
+}
+
+type LayerFactory func(layer uint) LayerReadWriter
```

Comment:
I thought the `Factory` part would make it clear, but since it's not really a factory I see where the confusion is coming from. Not sure what to name it, though.

`LayerProducer`? `LayerProducingFunction`? Better ideas?

Reply:
GenrateLayer (func names should usually be commands)

Reply:
I agree that **function** names should be commands. But my intuition was that it doesn't fit a delegate so I looked it up and the interwebs seems to [agree](https://stackoverflow.com/questions/2346065/proper-naming-convention-for-a-net-delegate-type) with me.

Classic example: an event handler delegate is usually named `somethingSomethingHandler`. The function of this type is then named `handleSomethingSomething`.

In our case, the function should be called `generateLayer` and be of type `LayerGenerator`.

I've changed the function name to `generateLayer`, but I've kept the delegate name, since a factory seems like a better than a generator. Let me know if you still disagree.